### PR TITLE
BERT nightly benchmark on Inferentia1

### PR DESCRIPTION
**Pull request by namannandan**
## Description
Benchmark BERT model on Inferentia1 instance

Model artifacts:
- https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_1.mar
- https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_2.mar
- https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_4.mar
- https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_8.mar

[Self hosted runner](https://github.com/pytorch/serve/settings/actions/runners/43)(inf1.6xlarge):
- 24 vCPUs
- 4 Inferentia1 chips (4 neuron cores per chip)

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

#### Checkpoint file generation
**Note: The artifacts above were traced using `transformers` version `4.6.0` as documented in the [Inferentia tutorial](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.html). With more recent `transformers` versions, the traced model for Neuron may generate incorrect inference result. Model output is `NaN`.**
```
$ cd examples/Huggingface_Transformers/
$ cat setup_config.json
{
 "model_name":"bert-base-uncased",
 "mode":"sequence_classification",
 "do_lower_case":true,
 "num_labels":"2",
 "save_mode":"torchscript",
 "max_length":"150",
 "captum_explanation":false,
 "embedding_name": "bert",
 "FasterTransformer":false,
 "BetterTransformer":false,
 "model_parallel":false,
 "hardware": "neuron",
 "batch_size": "2"
}
$ python Download_Transformer_models.py setup_config.json
$ ls Transformer_model/
traced_bert-base-uncased_model_neuron_batch_2.pt
```

#### MAR file generation
```
$ cat requirements.txt
torch-neuron
$ torch-model-archiver --model-name BERTSeqClassification_torchscript_neuron_batch_2 --version 1.0 --serialized-file ./examples/Huggingface_Transformers/Transformer_model/traced_bert-base-uncased_model_neuron_batch_2.pt --handler ./examples/Huggingface_Transformers/Transformer_handler_generalized_neuron.py --extra-files "./examples/Huggingface_Transformers/setup_config.json,./examples/Huggingface_Transformers/Seq_classification_artifacts/index_to_name.json,./examples/Huggingface_Transformers/Transformer_handler_generalized.py" --requirements-file requirements.txt
```

#### Benchmark run
```
$ cat benchmarks/benchmark_config_neuron.yaml
# Torchserve version is to be installed. It can be one of the options
#  - branch : "master"
#  - nightly: "2022.3.16"
#  - release: "0.5.3"
# Nightly build will be installed if "ts_version" is not specifiged
#ts_version:
#    branch: &ts_version "master"

# a list of model configure yaml files defined in benchmarks/models_config
# or a list of model configure yaml files with full path
models:
  - "bert_neuron_batch_2.yaml"

# benchmark on "cpu", "gpu" or "neuron".
# "cpu" is set if "hardware" is not specified
hardware: &hardware "neuron"
$
$ cat benchmarks/models_config/bert_neuron_batch_2.yaml
---
bert:
  scripted_mode:
    benchmark_engine: "ab"
    url: "file:///home/ubuntu/pytorch/model_store/BERTSeqClassification_torchscript_neuron_batch_2.mar"
    workers:
      - 4
    batch_delay: 100
    batch_size:
      - 2
    input: "./examples/Huggingface_Transformers/Seq_classification_artifacts/sample_text.txt"
    requests: 10000
    concurrency: 100
    backend_profiling: False
    exec_env: "local"
    processors:
      - "neuron"
$
$ python benchmarks/auto_benchmark.py --input benchmarks/benchmark_config_neuron.yaml
$ 
$ cat /tmp/ts_benchmark/scripted_mode_bert_w4_b2/ab_report.csv
Benchmark,Batch size,Batch delay,Workers,Model,Concurrency,Input,Requests,TS failed requests,TS throughput,TS latency P50,TS latency P90,TS latency P99,TS latency mean,TS error rate,Model_p50,Model_p90,Model_p99,predict_mean,handler_time_mean,waiting_time_mean,worker_thread_mean,cpu_percentage_mean,memory_percentage_mean,gpu_percentage_mean,gpu_memory_percentage_mean,gpu_memory_used_mean
AB,2,100,4,[.mar](file:///home/ubuntu/pytorch/model_store/BERTSeqClassification_torchscript_neuron_batch_2.mar),100,[input](./examples/Huggingface_Transformers/Seq_classification_artifacts/sample_text.txt),10000,0,436.38,225,234,254,229.157,0.0,17.14,18.07,22.93,17.44,17.31,207.2,0.27,0.0,0.0,0.0,0.0,0.0
```

#### Workflow test
Test branch: [`test-neuron-benchmark-workflow`](https://github.com/pytorch/serve/tree/test-neuron-benchmark-workflow)
Successful workflow run and artifacts: https://github.com/pytorch/serve/actions/runs/4352309159

Benchmark results:

TorchServe Benchmark on neuron
==============================

# Date: 2023-03-07 23:18:08

# TorchServe Version: torchserve-nightly==2023.3.6

## scripted_mode_bert_neuron_batch_1

|version|Benchmark|Batch size|Batch delay|Workers|Model|Concurrency|Input|Requests|TS failed requests|TS throughput|TS latency P50|TS latency P90|TS latency P99|TS latency mean|TS error rate|Model_p50|Model_p90|Model_p99|predict_mean|handler_time_mean|waiting_time_mean|worker_thread_mean|cpu_percentage_mean|memory_percentage_mean|gpu_percentage_mean|gpu_memory_percentage_mean|gpu_memory_used_mean|
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
|torchserve-nightly==2023.3.6|AB|1|100|4|[.mar](https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_1.mar)|100|[input](./examples/Huggingface_Transformers/Seq_classification_artifacts/sample_text.txt)|10000|0|196.42|505|601|678|509.115|0.0|14.31|39.43|51.18|19.52|19.42|485.66|0.23|0.0|8.5|0.0|0.0|0.0|

## scripted_mode_bert_neuron_batch_2

|version|Benchmark|Batch size|Batch delay|Workers|Model|Concurrency|Input|Requests|TS failed requests|TS throughput|TS latency P50|TS latency P90|TS latency P99|TS latency mean|TS error rate|Model_p50|Model_p90|Model_p99|predict_mean|handler_time_mean|waiting_time_mean|worker_thread_mean|cpu_percentage_mean|memory_percentage_mean|gpu_percentage_mean|gpu_memory_percentage_mean|gpu_memory_used_mean|
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
|torchserve-nightly==2023.3.6|AB|2|100|4|[.mar](https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_2.mar)|100|[input](./examples/Huggingface_Transformers/Seq_classification_artifacts/sample_text.txt)|10000|0|575.7|163|198|230|173.702|0.0|12.28|14.42|34.95|13.15|13.06|156.46|0.18|0.0|0.0|0.0|0.0|0.0|

## scripted_mode_bert_neuron_batch_4

|version|Benchmark|Batch size|Batch delay|Workers|Model|Concurrency|Input|Requests|TS failed requests|TS throughput|TS latency P50|TS latency P90|TS latency P99|TS latency mean|TS error rate|Model_p50|Model_p90|Model_p99|predict_mean|handler_time_mean|waiting_time_mean|worker_thread_mean|cpu_percentage_mean|memory_percentage_mean|gpu_percentage_mean|gpu_memory_percentage_mean|gpu_memory_used_mean|
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
|torchserve-nightly==2023.3.6|AB|4|100|4|[.mar](https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_4.mar)|100|[input](./examples/Huggingface_Transformers/Seq_classification_artifacts/sample_text.txt)|10000|0|652.79|149|150|240|153.188|0.0|22.97|23.16|23.65|23.12|23.02|125.6|0.48|0.0|0.0|0.0|0.0|0.0|

## scripted_mode_bert_neuron_batch_8

|version|Benchmark|Batch size|Batch delay|Workers|Model|Concurrency|Input|Requests|TS failed requests|TS throughput|TS latency P50|TS latency P90|TS latency P99|TS latency mean|TS error rate|Model_p50|Model_p90|Model_p99|predict_mean|handler_time_mean|waiting_time_mean|worker_thread_mean|cpu_percentage_mean|memory_percentage_mean|gpu_percentage_mean|gpu_memory_percentage_mean|gpu_memory_used_mean|
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
|torchserve-nightly==2023.3.6|AB|8|100|4|[.mar](https://torchserve.pytorch.org/mar_files/BERTSeqClassification_torchscript_neuron_batch_8.mar)|100|[input](./examples/Huggingface_Transformers/Seq_classification_artifacts/sample_text.txt)|10000|0|649.43|150|164|171|153.98|0.0|46.95|47.26|48.51|47.05|46.93|101.99|0.48|0.0|0.0|0.0|0.0|0.0|

#### Consolidated benchmark workflow test
Test branch: [`test-neuron-benchmark-workflow`](https://github.com/pytorch/serve/tree/test-neuron-benchmark-workflow)
Successful workflow run: https://github.com/pytorch/serve/actions/runs/4400212613

#### Regression test
CPU: [neuron_benchmark_regression_log_cpu.txt](https://github.com/pytorch/serve/files/10962229/neuron_benchmark_regression_log_cpu.txt)
GPU: [neuron_benchmark_regression_log_gpu.txt](https://github.com/pytorch/serve/files/10962233/neuron_benchmark_regression_log_gpu.txt)

## Checklist:

- [x] Did you have fun?
- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by lxning** [comment]
Please file a ticket to inferentia about transformer==4.6.0 issue.

**Review by agunapal — COMMENTED** [review]
@namannandan If we are going to run this nightly, It would be good to just add it to the benchmark_gpu workflow and modify it to run on both the machines with the matrix command. The yaml file can be changed with an if else statement.

**Comment by namannandan** [comment]
@lxning Inferentia team now has an internal ticket to track the issue with `torch-neuron` library being unable to correctly trace models when `transformers` package version is `>4.19`. Currently, the model archives I've prepared have checkpoint files that were traced using `transformers` version `4.6.0`. I tested that the inference with these model archives works as expected with the latest transformers package version as of this writing which is `4.26.1`. So the issue is only with tracing the model. We should be able to go ahed and start running the `Inf1` benchmark and once the Inferentia team fixes the issue with `torch-neuron` we can just re-create the model archives and upload them to the model zoo. Please let me know your thoughts.

@agunapal since these models are traced and intended to run on Inferentia1 which is an additional hardware platform alongside CPU and GPU, would it make sense to maintain this in a separate workflow? Unless there is a downside to creating separate workflows for different hardware platforms. Please let me know your thoughts.

**Comment by lxning** [comment]
@namannandan you can test it the workflows you created in the PR and link the result at here.

**Comment by namannandan** [comment]
@lxning tested workflow and linked benchmark results in the PR summary above.

Based on offline discussion with @agunapal, consolidation of workflow files for benchmarking using matrix command to run on different runners makes sense. This can be implemented in a separate PR.

**Comment by lxning** [comment]
@namannandan 
- can you run regression test to make sure there are no breaks for the existing test for hf transformers.
- as @ankithagunapal suggested, update github workflow by using matrix in this PR so that one single PR can cover everything of this task.
Thanks a lot.

**Comment by namannandan** [comment]
@lxning, @agunapal
- Logs for regression tests
CPU: [neuron_benchmark_regression_log_cpu.txt](https://github.com/pytorch/serve/files/10962229/neuron_benchmark_regression_log_cpu.txt)
GPU: [neuron_benchmark_regression_log_gpu.txt](https://github.com/pytorch/serve/files/10962233/neuron_benchmark_regression_log_gpu.txt)
- Consolidated benchmark workflow files into a single file. Successful workflow run and artifacts: https://github.com/pytorch/serve/actions/runs/4352309159

File: .github/workflows/benchmark_nightly.yml [hunk 1]
```diff
@@ -30,18 +35,39 @@ jobs:
...
         uses: actions/checkout@v3
-      - name: Install dependencies
+      - name: Install dependencies cpu
+        if: ${{ matrix.hardware == 'cpu' }}
         run: |
-            sudo apt-get update -y
-            sudo apt-get install -y apache2-utils
-            pip install -r benchmarks/requirements-ab.txt
-            export omp_num_threads=1
+          sudo apt-get update -y
```
[the remark this answers is not in the document]
Comment (min-jean-cho) [reply .github/workflows/benchmark_nightly.yml:41]: Yes, currently it's only applicable to CPU benchmarks. By the way, I noticed `export omp_num_threads=1` doesn't correctly set `OMP_NUM_THREADS` to `1`, https://github.com/pytorch/serve/pull/2151 You may want to double check in the github action.

[the remark this answers is not in the document]
Comment (namannandan) [reply .github/workflows/benchmark_nightly.yml:41]: @agunapal makes sense, I'll update the install dependencies step.

@min-jean-cho thanks for spotting that! I'll fix the env var.

[the remark this answers is not in the document]
Comment (min-jean-cho) [reply .github/workflows/benchmark_nightly.yml:41]: Thanks @namannandan, by the way I recall setting the environment variable `export OMP_NUM_THREADS=1` seemed not to be correctly setting the number of threads, https://github.com/pytorch/serve/pull/2151/files#diff-cac3a24029ba9498c7e1735f8fc6e65b5a8a090d7f015bd3c35051f57a9981caR178-R179. You may also want to have a double check, thanks!

[the remark this answers is not in the document]
Comment (namannandan) [reply .github/workflows/benchmark_nightly.yml:41]: Ah I see, I'll double check that. I wonder using the `env` key in the `Benchmark cpu nightly` step would do the trick. It is documented [here](https://docs.github.com/en/actions/learn-github-actions/variables#defining-environment-variables-for-a-single-workflow). I'll try this method as well.

[the remark this answers is not in the document]
Comment (namannandan) [reply .github/workflows/benchmark_nightly.yml:41]: Looks like using the `env` key works as expected:
https://github.com/pytorch/serve/actions/runs/4410687740/workflow#L48

CPU benchmark logs:
```
OMP_NUM_THREADS:  1
torch.get_num_threads:  1
NEURON_RT_NUM_CORES: 
```

Inf1 benchmark logs:
```
OMP_NUM_THREADS: 
torch.get_num_threads:  12
NEURON_RT_NUM_CORES:  4
```

File: examples/Huggingface_Transformers/README.md [hunk 2]
```diff
@@ -51,21 +51,25 @@ In the setup_config.json :
 
 *embedding_name* : The name of embedding layer in the chosen model, this could be `bert` for `bert-base-uncased`, `roberta` for `roberta-base` or `roberta` for `xlm-roberta-large`, or `gpt2` for `gpt2` model
 
+*hardware* : The target platform to trace the model for. Specify as `neuron` for Inferentia1.
+
+*batch_size* : Input batch size when tracing the model for `neuron` as target hardware.
```
Comment (agunapal) [inline examples/Huggingface_Transformers/README.md:56]: Is this applicable only to Inferentia?

Diff looks good overall.

Nit: Please mention  a sentence or two about Inferentia or point to a url with details.

Also, do we want to mention about this in the benchmark ReadMe

Comment (namannandan) [reply examples/Huggingface_Transformers/README.md:56]: Yes, that's correct. Currently `batch_size` only affects model tracing for Inferentia.
Will update docs to include reference to Inferentia.

**Review by agunapal — APPROVED** [review]
LGTM

**Comment by namannandan** [comment]
Successful consolidated benchmark workflow run: https://github.com/pytorch/serve/actions/runs/4451500075

**Review by msaroufim — APPROVED** [review]

**Pull request merged**